OpenRA.Mods.Cnc/UtilityCommands/ImportGen2MapCommand.cs

```diff
@@ -410,85 +163,28 @@ static void ReadOverlay(Map map, IniFile file, int2 fullSize)
 				if (overlayType == 0xFF)
 					continue;
 
-				if (OverlayToActor.TryGetValue(overlayType, out var actorType))
+				if (TryHandleOverlayToActorInner(cell, overlayPack, overlayIndex, overlayType, out var ar))
 				{
-					if (string.IsNullOrEmpty(actorType))
-						continue;
-
-					var shape = new Size(1, 1);
-					if (OverlayShapes.TryGetValue(overlayType, out shape))
-					{
-						// Only import the top-left cell of multi-celled overlays
-						var aboveType = overlayPack[overlayIndex[cell - new CVec(1, 0)]];
-						if (shape.Width > 1 && aboveType != 0xFF)
-							if (OverlayToActor.TryGetValue(aboveType, out var a) && a == actorType)
-								continue;
-
-						var leftType = overlayPack[overlayIndex[cell - new CVec(0, 1)]];
-						if (shape.Height > 1 && leftType != 0xFF)
-							if (OverlayToActor.TryGetValue(leftType, out var a) && a == actorType)
-								continue;
-					}
-
-					// Fix position of vein hole actors
-					var location = cell;
-					if (actorType == "veinhole")
-						location -= new CVec(1, 1);
-
-					var ar = new ActorReference(actorType)
-					{
-						new LocationInit(location),
-						new OwnerInit("Neutral")
-					};
-
-					if (OverlayToHealth.TryGetValue(overlayType, out var damageState))
-					{
-						var health = 100;
-						if (damageState == DamageState.Critical)
-							health = 25;
-						else if (damageState == DamageState.Heavy)
-							health = 50;
-						else if (damageState == DamageState.Medium)
-							health = 75;
-
-						if (health != 100)
-							ar.Add(new HealthInit(health));
-					}
-
-					map.ActorDefinitions.Add(new MiniYamlNode("Actor" + map.ActorDefinitions.Count, ar.Save()));
+					if (ar != null)
+						map.ActorDefinitions.Add(new MiniYamlNode("Actor" + map.ActorDefinitions.Count, ar.Save()));
 
 					continue;
 				}
 
-				// TS maps encode the non-harvestable border tiles as overlay
-				// Only convert to vein resources if the overlay data specifies non-border frames
-				if (overlayType == 0x7E)
+				if (TryHandleResourceFromOverlayInner(overlayType, out var resourceType))
```

Comment:
IMO this should be returning the `ResourceTile` in case implementations want to modify the density.